File: Commons/Data/src/main/java/com/greentree/commons/data/resource/location/CacheResourceLocation.java
package com.greentree.commons.data.resource.location;

import com.greentree.commons.data.resource.IOResource;
import com.greentree.commons.data.resource.Resource;
import com.greentree.commons.data.resource.ResourceAction;

import java.io.InputStream;
import java.io.OutputStream;
import java.net.MalformedURLException;
import java.net.URL;

public class CacheResourceLocation implements IOResourceLocation {

    private static final long serialVersionUID = 1L;

    private final ResourceLocation sourceLocation;

    private final IOResourceLocation cacheLocation;

    public CacheResourceLocation(ResourceLocation sourceLocation,
                                 IOResourceLocation cacheLocation) {
        this.sourceLocation = sourceLocation;
        this.cacheLocation = cacheLocation;
    }

    @Override
    public void clear() {
        cacheLocation.clear();
    }

    @Override
    public synchronized IOResource getResource(String name) {
        final var cacheName = name;
        if (cacheLocation.isExist(cacheName)) {
            if (sourceLocation.isExist(name)) {
                final var cache = cacheLocation.getResource(cacheName);
                final var source = sourceLocation.getResource(name);
                cacheLocation.deleteResource(cacheName);
                source.writeTo(cache, cache.lastModified());
                return new CacheResource(source, cache);
            } else {
                cacheLocation.deleteResource(cacheName);
                return null;
            }
        } else if (sourceLocation.isExist(name)) {
            final var cache = cacheLocation.createResource(cacheName);
            final var source = sourceLocation.getResource(name);
            source.writeTo(cache);
            return new CacheResource(source, cache);
        } else
            return null;
    }

    @Override
    public IOResource createNewResource(String name) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean deleteResource(String name) {
        return cacheLocation.deleteResource(name);
    }

    public static final class CacheResource implements IOResource {

        private static final long serialVersionUID = 1L;
        private final Resource source;
        private final IOResource cache;

        public CacheResource(Resource source, IOResource cache) {
            this.source = source;
            this.cache = cache;
        }

        @Override
        public ResourceAction getAction() {
            return source.getAction();
        }

        @Override
        public String getName() {
            return cache.getName();
        }

        @Override
        public long length() {
            return cache.length();
        }

        @Override
        public long lastModified() {
            return cache.lastModified();
        }

        @Override
        public InputStream open() {
            return cache.open();
        }

        @Override
        public URL url() throws MalformedURLException {
            return cache.url();
        }

        @Override
        public boolean delete() {
            return cache.delete();
        }

        @Override
        public boolean exists() {
            return cache.exists();
        }

        @Override
        public OutputStream openWrite() {
            throw new UnsupportedOperationException();
        }

    }

}

File: Commons/Data/src/main/java/com/greentree/commons/data/resource/location/CacheResourceLocation.kt
package com.greentree.commons.data.resource.location

import com.greentree.commons.data.resource.FileResource
import com.greentree.commons.data.resource.MutableFileResource
import com.greentree.commons.data.resource.ParentResource
import com.greentree.commons.data.resource.writeTo
import java.io.InputStream

class CacheResourceLocation(
	private val sourceLocation: ResourceLocation,
	private val cacheLocation: MutableResourceLocation,
) : ResourceLocation {

	override fun getResourceOrNull(name: String): FileResource? {
		val source = sourceLocation.getFileResourceOrNull(name) ?: return null
		val cache = cacheLocation.getFileResource(name)
		return CacheFileResource(source, cache)
	}

	override fun getFileResourceOrNull(name: String): FileResource? {
		val source = sourceLocation.getFileResourceOrNull(name) ?: return null
		val cache = cacheLocation.getFileResource(name)
		return CacheFileResource(source, cache)
	}

	data class CacheFileResource(private val source: FileResource, private val cache: MutableFileResource) :
		FileResource {

		override fun exists() = source.exists()

		override val length: Long
			get() {
				update()
				return source.length
			}

		override fun open(): InputStream {
			update()
			return cache.open()
		}

		override val parent: ParentResource
			get() = TODO("Not yet implemented")
		override val name: String
			get() = source.name

		override fun lastModified() = source.lastModified()

		private fun update() {
			source.writeTo(cache, cache.lastModified())
		}
	}
}